src/states/node.rs

```diff
@@ -809,6 +832,143 @@ impl Node {
         }
     }
 
+    fn handle_node_approval_vote(&mut self,
+                                 candidate_name: XorName,
+                                 validity: bool)
+                                 -> Evented<Result<(), RoutingError>> {
+        let mut result = Evented::empty();
+        // Once the joining node joined, it may receive the vote regarding itself.
+        // Or a node may receive CandidateApproval before connection established.
+        let (client_auth, peer_id) = try_ev!(
+            self.peer_mgr.handle_node_approval_vote(candidate_name, validity),
+            result);
+
+        if validity {
+            let sections = self.peer_mgr.get_sections(self.full_id.public_id());
+            info!("{:?} Sending NodeApproval to {:?}.", self, candidate_name);
+            let _ = self.send_routing_message(RoutingMessage {
+                    src: Authority::Section(candidate_name),
+                    dst: client_auth,
+                    content: MessageContent::NodeApproval { sections: sections },
+                })
+                .extract(&mut result);
+        } else {
+            self.disconnect_peer(&peer_id);
+        }
+        result.map(Ok)
+    }
+
+    fn handle_approval_confirmation(&mut self,
+                                    candidate_name: XorName)
+                                    -> Evented<Result<(), RoutingError>> {
+        let mut result = Evented::empty();
+        let (pub_id, peer_id) = try_ev!(self.peer_mgr.handle_approval_confirmation(candidate_name),
+                                        result);
+        self.add_to_routing_table(&pub_id, &peer_id).extract(&mut result);
+        result.map(Ok)
+    }
+
+    fn handle_node_approval(&mut self,
+                            groups: Vec<(Prefix<XorName>, Vec<PublicId>)>)
+                            -> Evented<Result<(), RoutingError>> {
+        let mut events = Evented::empty();
+        if self.is_approved {
+            warn!("{:?} Received duplicate NodeApproval.", self);
+            return events.with_value(Ok(()));
+        }
+
+        self.peer_mgr.populate_routing_table(&groups);
```

Comment:
`populate_routing_table` should probably be called `update_prefixes`, since it doesn't really add any nodes. And it should take an `IntoIterator<item = Prefix>`, I think. It's misleading to pass the whole groups into it if the get ignored. This call should look like this, I think:
```rust
self.peer_mgr.update_prefixes(groups.iter().map(|&(ref prefix, _)| prefix));
```